src/main/java/ch/uzh/ifi/hase/soprafs22/service/GameService.java

```diff
@@ -243,10 +250,72 @@ public Game getGameFromRandomUser(User user) {
     PageRequest pageRequest = PageRequest.of(pageIndex, 1);
 
     // get the page with the game
-    Page<Game> somePage = gameRepository.getOtherUserWithActiveGameThatWasNotPlayedOn(pageRequest, user.getId(), user,
-        user.getGender(), minAgeDate, maxAgeDate, user.getBlockedUsers(), user.getMatchedUsers());
+    Page<Game> somePage = gameRepository.getOtherUserWithActiveGameThatWasNotPlayedOn(
+      pageRequest, 
+      user.getId(), 
+      user,
+      user.getGender(), 
+      minAgeDate, 
+      maxAgeDate, 
+      user.getBlockedUsers(),
+      user.getMatchedUsers()
+    );
+
+    // subset the page of users to only users that have a haversine distance of less than user.getMaxRange()
+    List<Game> games = somePage.getContent();
+    List<Game> gamesWithDistance = new ArrayList<>();
+    // go over all games that match the other criteria
+    for(Game game : games){
+      // for each game get the user that the game belongs to
+      User gameUser = game.getUser();
+
+      // calculate the distance between the two users
+      double distance = haversineDistance(
+        user.getLatitude(), 
+        user.getLongitude(), 
+        gameUser.getLatitude(), 
+        gameUser.getLongitude()
+      );
+
+      // QUESTION TO REVIEWER: right now i compute the distance for every user i find that matches the other preference criteria. However, I only need to compute it until I find one that has a distance that is 
+
+      // if the distance is less than the max range, add the game to the list
+      if(distance <= user.getMaxRange()){
+        gamesWithDistance.add(game);
+      }
+    }
 
     // return the game
-    return somePage.getContent().get(0);
+    return gamesWithDistance.get(0);
```

Comment:
It is possible that at this point, gameWithDistance is empty, so this will throw an error.. We could use somePage.getContent().get(0) only because we first counted the valid games, so that it is garantueed to return a game (else we would throw 404 at the count above). IMO the entire function might need some refactoring and overthinkig.